File: checks/python/src/section508/dom.py
"""Small BeautifulSoup helpers that mirror the JS fragment-cop DOM calls."""

from __future__ import annotations

import copy
import math
import re

import soupsieve
from bs4 import NavigableString, Tag

SPACE = re.compile(r"\s+")


def is_tag(node) -> bool:
    return isinstance(node, Tag)


def tag_name(node) -> str:
    if not is_tag(node) or not node.name:
        return ""
    return str(node.name).lower()


def select(root, selector: str) -> list[Tag]:
    if root is None or not hasattr(root, "select"):
        return []
    try:
        return list(root.select(selector))
    except (ValueError, NotImplementedError, soupsieve.SelectorSyntaxError):
        return []


def matches(node, selector: str) -> bool:
    if not is_tag(node) or tag_name(node) in {"", "[document]"}:
        return False
    try:
        return bool(soupsieve.match(selector, node))
    except (ValueError, NotImplementedError, soupsieve.SelectorSyntaxError):
        return False


def inclusive_select(root, selector: str) -> list[Tag]:
    found = []
    if matches(root, selector):
        found.append(root)
    found.extend(select(root, selector))
    return found


def closest(node, selector: str) -> Tag | None:
    current = node if is_tag(node) else None
    while current is not None and is_tag(current) and tag_name(current) != "[document]":
        if matches(current, selector):
            return current
        current = current.parent
    return None


def has_attr(node, name: str) -> bool:
    return is_tag(node) and node.has_attr(name)


def attr(node, name: str):
    """Like getAttribute: None if missing, else a string (possibly empty)."""
    if not has_attr(node, name):
        return None
    value = node.get(name)
    if value is None or value is True:
        return ""
    if isinstance(value, list):
        return " ".join(str(item) for item in value)
    return str(value)


def attr_or(node, name: str, default: str = "") -> str:
    value = attr(node, name)
    return default if value is None else value


def text_of(node) -> str:
    if node is None:
        return ""
    if isinstance(node, NavigableString):
        return str(node)
    if is_tag(node):
        return node.get_text()
    return str(node)


def norm_text(node) -> str:
    return SPACE.sub(" ", text_of(node)).strip()


def class_string(node) -> str:
    return attr_or(node, "class")


def has_class(node, name: str) -> bool:
    return name in class_string(node).split()


def children_tags(node) -> list[Tag]:
    if not is_tag(node):
        return []
    return [child for child in node.children if is_tag(child)]


def significant_children(node) -> list:
    if not is_tag(node):
        return []
    out = []
    for child in node.children:
        if isinstance(child, NavigableString):
            if str(child).strip():
                out.append(child)
        elif is_tag(child):
            out.append(child)
    return out


def clone(node: Tag) -> Tag:
    return copy.copy(node)


def remove_matches(node: Tag, selector: str) -> None:
    for child in list(select(node, selector)):
        child.decompose()


def find_id(root, id_value: str) -> Tag | None:
    if not id_value:
        return None
    if is_tag(root) and attr(root, "id") == id_value:
        return root
    if root is None or not hasattr(root, "find"):
        return None
    return root.find(id=id_value)


def label_for(root, id_value: str) -> Tag | None:
    if not id_value or root is None or not hasattr(root, "find"):
        return None
    return root.find("label", attrs={"for": id_value})


def parse_number(value) -> float | None:
    if value is None:
        return None
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    if not math.isfinite(number):
        return None
    return number

File: checks/ruby/lib/section508/dom.rb
# frozen_string_literal: true

module Section508
  # Small DOM helpers so cops can follow the JS querySelector style on Nokogiri.
  module Dom
    module_function

    def element?(node)
      node.is_a?(Nokogiri::XML::Element)
    end

    def elements(root, selector)
      return [] unless root.respond_to?(:css)

      root.css(selector).select { |node| element?(node) }
    end

    def first(root, selector)
      elements(root, selector).first
    end

    def elements_including_self(root, selector)
      list = []
      list << root if matches?(root, selector)
      list.concat(elements(root, selector))
      list
    end

    def matches?(node, selector)
      return false unless element?(node)

      if node.respond_to?(:matches?)
        node.matches?(selector)
      else
        parent = node.parent
        parent && parent.respond_to?(:css) && parent.css(selector).include?(node)
      end
    rescue Nokogiri::CSS::SyntaxError, ArgumentError
      false
    end

    def closest(node, selector)
      current = node
      while element?(current)
        return current if matches?(current, selector)

        current = current.parent
      end
      nil
    end

    def parent_element(node)
      parent = node.parent
      parent if element?(parent)
    end

    def tag(node)
      node.name.to_s.upcase
    end

    def attr(node, name)
      node[name.to_s]
    end

    def has_attr?(node, name)
      node.key?(name.to_s)
    end

    def has_class?(node, name)
      node["class"].to_s.split.include?(name)
    end

    def text(node)
      node.text.to_s
    end

    def norm(node)
      text(node).gsub(/\s+/, " ").strip
    end

    def own_text(node, remove: "script, style")
      clone = node.dup
      clone.css(remove).each(&:remove) if remove
      clone.text.to_s.gsub(/\s+/, " ").strip
    end

    def by_id(root, id)
      return nil if id.nil? || id.to_s.empty?
      return root if element?(root) && root["id"] == id

      elements(root, "[id]").find { |node| node["id"] == id }
    end

    def label_for(root, id)
      return nil if id.nil? || id.to_s.empty?

      elements(root, "label[for]").find { |node| node["for"] == id }
    end

    def associated_label_text(control, root)
      wrap = closest(control, "label")
      wrapped = wrap && label_text(wrap, control)
      return wrapped if wrapped && !wrapped.empty?

      id = control["id"]
      return nil if id.nil? || id.empty?

      label = label_for(root, id)
      return nil unless label

      text = norm(label)
      text unless text.empty?
    end

    def has_associated_visible_label?(control, root)
      !associated_label_text(control, root).to_s.empty?
    end

    def label_text(label, _control)
      clone = label.dup
      clone.css("input, select, textarea, button").each(&:remove)
      clone.text.to_s.gsub(/\s+/, " ").strip
    end

    def field_hint(control)
      type = (control["type"] || control.name.to_s).to_s.gsub(/\s+/, " ")
      name = (control["name"] || control["id"] || type).to_s.strip
      name.empty? ? "this field" : name
    end

    def described_text(element, root)
      ids = element["aria-describedby"].to_s.split(/\s+/).reject(&:empty?)
      ids.map { |id| by_id(root, id) }.compact.map { |node| text(node) }.join(" ")
    end

    def labelledby_text(element, root)
      ids = element["aria-labelledby"].to_s.split(/\s+/).reject(&:empty?)
      ids.map { |id| by_id(root, id) }.compact.map { |node| text(node) }.join(" ").gsub(/\s+/, " ").strip
    end
  end
end
